### PR TITLE
Rename `profiler` directory to `profilers`

### LINKED ISSUE #11972: Rename `profiler/` directory to `profilers/`
## Proposed refactor

Rename `profiler/` directory to `profilers/`

### Motivation

The other directory names are plural, like `callbacks/`, `strategies/`, `accelerators/`- let's do the same for profilers. Part of https://github.com/PyTorchLightning/pytorch-lightning/issues/11908

cc @borda @justusschock @awaelchli @rohitgr7 @carmocca @kaushikb11 @ninginthecloud

### PR DESCRIPTION
**Pull request by Pragyanstha**
# [rebase and merge]
## What does this PR do?
Closes #11972.

- Consists of two commits:
  - Commit 1:
    - rename `profiler/` directory to `profilers/`.
    - update import statements all over the repository.
  - Commit 2:
    - reroute `profiler/` to `profilers/` with the deprecation warning.
    - put back items that had already been deprecated in the previous PRs.

### Does your PR introduce any breaking changes? If yes, please list them.
No, it just deprecates `profiler/*` in favour of `profilers/*`.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
maybe

### CODE REVIEW

File: pytorch_lightning/profilers/__init__.py
Comment (rohitgr7) [inline pytorch_lightning/profilers/__init__.py:1]: can you add the license from another module here?

**Review by rohitgr7 — APPROVED** [review]
LGTM! mind add a changelog entry?

File: pytorch_lightning/profilers/__init__.py
Comment (Pragyanstha) [reply pytorch_lightning/profilers/__init__.py:1]: sure

**Comment by kaushikb11** [comment]
> A second PR would be created for adding back the previous API that redirects the call into the new API,
with a deprecation warning.

I think you should add it in this PR

**Comment by rohitgr7** [comment]
> > A second PR would be created for adding back the previous API that redirects the call into the new API,
> > with a deprecation warning.
> 
> I think you should add it in this PR

it will create a lot of changes in a single one since git file-renames won't be valid then.

File: .github/CODEOWNERS
Comment (carmocca) [inline .github/CODEOWNERS:34]:
```suggestion
/pytorch_lightning/profilers            @williamfalcon @tchaton @borda @carmocca
/pytorch_lightning/profilers/pytorch.py @nbcsm @guotuofeng
```

**Review by carmocca — APPROVED** [review]
LGTM but this should go after the release to make sure we have time for the deprecation too.

**Review by tchaton — APPROVED** [review]
LGTM !

**Comment by daniellepintz** [comment]
@Pragyanstha mind rebasing this PR?

**Review by ananthsub — CHANGES_REQUESTED** [review]
As is, this PR is a breaking change for those on master, which we should avoid.

I would strongly recommend doing both of these in the same PR to avoid any breaking change:
- Copy the existing code over to a profilers/ directory
- In the existing profiler/ directory, import the components from the new profilers/ directory and include the relevant deprecation warnings

**Comment by rohitgr7** [comment]
> As is, this PR is a breaking change for those on master, which we should avoid.
> 
> I would strongly recommend doing both of these in the same PR to avoid any breaking change:
> 
>     * Copy the existing code over to a profilers/ directory
> 
>     * In the existing profiler/ directory, import the components from the new profilers/ directory and include the relevant deprecation warnings

wouldn't that create a huge git diff? I think we did the same for `training type plugin -> strategy`. This is targeted after v1.6, so we can quickly create another one once this is merged.

**Comment by akihironitta** [comment]
I'll clean up the commits and have this PR ready for "rebase and merge" instead of for "squash and merge" similarly to #12740 and #13031.

File: pytorch_lightning/trainer/trainer.py
Comment (akihironitta) [inline pytorch_lightning/trainer/trainer.py:174]: @carmocca Should I undo the change here for someone to address it in a different PR (#13046)? This PR needs it fixed for merge.

Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:174]: If this PR gets merged before it is fixed in the other PR, we can still apply the fix to the bugfix branch directly

**Comment by akihironitta** [comment]
Dismissing all reviews as a lot of changes have been introduced since then.

**Review by carmocca — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
Please note, https://github.com/PyTorchLightning/pytorch-lightning/pull/13084 has changed a type hint. This change should be carried over as well.

**Review by awaelchli — APPROVED** [review]
I rebased this to the latest code structure changes

**Comment by carmocca** [comment]
@awaelchli This PR will be "Rebase and Merged" since it's moving and modifying code. So this needs to be rebased on top of master with only 2 clean commits.

**Review by Borda — APPROVED** [review]

File: CHANGELOG.md
Comment (carmocca) [inline CHANGELOG.md:140]: The commits are not correct: https://github.com/Lightning-AI/lightning/pull/12308/commits/29d6cdaad861f533e9604bbd75ec73ea3b856177 should only be modifying the filenames (breaking compatibility) and the next one just adding back compatibility for the old name

Comment (awaelchli) [reply CHANGELOG.md:140]: What am I supposed to do

**Comment by carmocca** [comment]
Included a change in the last commit that should fix the cli failures

**Pull request merged**
